### PR TITLE
Refactor feature extraction key

### PR DESCRIPTION
**Pull request by ashwinvaidya17**
# Description

- All backbone/feature extraction keys are now in
```yaml
model:
    feture_extractor:
        backbone:
        layers:
````
- `FeatureExtractor` selects timm or torchfx based on the parameters passed to it.
- This PR ended up way bigger than I thought 😃 

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [x] My code follows the [pre-commit style and check guidelines](https://github.com/openvinotoolkit/anomalib/blob/main/CONTRIBUTING.md) of this project.
- [ ] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: anomalib/models/cflow/lightning_model.py [hunk 1]
```diff
@@ -30,9 +34,7 @@ class Cflow(AnomalyModule):
     def __init__(
         self,
         input_size: Tuple[int, int],
-        backbone: str,
-        layers: List[str],
-        pre_trained: bool = True,
+        feature_extractor: Union[TimmFeatureExtractorParams, TorchFXFeatureExtractorParams],
```
Comment (jpcbertoldo) [inline anomalib/models/cflow/lightning_model.py:37]: a minor suggestion (but it helps to keep things consistent as the code scales): there could be something like

```python
FeatureExtractorParamsTypes = Union[TimmFeatureExtractorParams, TorchFXFeatureExtractorParams]
```

to avoid this double-import-and-union

File: anomalib/models/components/feature_extractors/__init__.py [hunk 2]
```diff
@@ -3,14 +3,142 @@
...
+
+import torch
+from omegaconf import DictConfig
+from torch import Tensor, nn
+
+from .timm import TimmFeatureExtractor, TimmFeatureExtractorParams
+from .torchfx import TorchFXFeatureExtractor, TorchFXFeatureExtractorParams
+
+
+class FeatureExtractor(nn.Module):
```
Comment (jpcbertoldo) [inline anomalib/models/components/feature_extractors/__init__.py:16]: i find it weird to have behavior/interfaces defined in `__init__.py` 
idk if there is a practical advantage/disavantage but i don't recall to have seen this in other projects

**Review by jpcbertoldo — COMMENTED** [review]
great effort, thanks!

File: anomalib/models/cflow/lightning_model.py (hunk 1 above)
Comment (ashwinvaidya17) [reply anomalib/models/cflow/lightning_model.py:37]: That's a good idea

File: anomalib/models/components/feature_extractors/feature_extractor.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            Union[TimmFeatureExtractorParams, TorchFXFeatureExtractorParams]: Converted feature extractor parameters.
+        """
+        if isinstance(feature_extractor_params, (DictConfig, dict)):
+            if "layers" in feature_extractor_params:
+                feature_extractor_params = TimmFeatureExtractorParams(**feature_extractor_params)
+            else:
+                feature_extractor_params = TorchFXFeatureExtractorParams(**feature_extractor_params)
+        if not isinstance(feature_extractor_params, (TimmFeatureExtractorParams, TorchFXFeatureExtractorParams)):
+            raise ValueError(f"Unknown feature extractor params type: {type(feature_extractor_params)}")
+        return feature_extractor_params
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/feature_extractor.py:116]: If I understand correctly, the mechanism to distinguish between TorchFX and Timm depends on the occurrence of the `layers` key in the dictionary. I'm not sure if this is desirable. It might lead to confusing situations where the user thinks they're using torchvision backbone but actually it's Timm. 

How about setting the framework explicitly, by passing a `framework` parameter that defaults to `torchvision`?

File: anomalib/models/components/feature_extractors/feature_extractor.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+import torch
+from omegaconf import DictConfig
+from torch import Tensor, nn
+
+from .timm import TimmFeatureExtractor, TimmFeatureExtractorParams
+from .torchfx import TorchFXFeatureExtractor, TorchFXFeatureExtractorParams
+
+
+class FeatureExtractor(nn.Module):
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/feature_extractor.py:16]: I have some concerns about the class structure of the feature extractors, mainly caused by the design which involves several layers of wrappers around the feature extractor network.

I feel there are too many objects living in the torch model which are all referred to as 'feature extractor'. First of all, the torch model has a constructor argument named `feature extractor`, which is used to construct the `FeatureExtractor` class, which is then assigned to `self.feature_extractor`. 

```
self.feature_extractor = FeatureExtractor(feature_extractor)
```
This already looks a bit confusing.

Then, due to the wrapper design, this `self.feature_extractor` object now contains several layers of attributes which are all called `feature_extractor`. As a result, the actual network that performs the feature extraction resides at
```
self.feature_extractor.feature_extractor.feature_extractor
```
This makes the code around the feature extraction functionality a bit confusing and hard to navigate.

----
While I understand the motivation behind the design, I wonder if there might be a way to achieve the same thing without relying on this many wrappers. How about using a base `FeatureExtractor` class which has `TorchFXFeatureExtractor` and `TimmFeatureExtractor` as subclasses.

**Review by djdameln — COMMENTED** [review]
Thanks. I had a close look at the design and I have some concerns, which are outlined below.

File: anomalib/models/cflow/lightning_model.py (hunk 1 above)
Comment (jpcbertoldo) [reply anomalib/models/cflow/lightning_model.py:37]: I think it's nice good practice to follow and cheap to implement.
Maybe an exception in mypy checks?

File: anomalib/models/dfkde/torch_model.py [hunk 5]
```diff
@@ -48,7 +49,7 @@ def __init__(
         self.threshold_steepness = threshold_steepness
         self.threshold_offset = threshold_offset
 
-        self.feature_extractor = FeatureExtractor(backbone=backbone, pre_trained=pre_trained, layers=layers).eval()
+        self.feature_extractor = FeatureExtractor(feature_extractor).eval()
```
Comment (djdameln) [inline anomalib/models/dfkde/torch_model.py:52]: Would it be an idea to set the feature extractor modules to eval mode by default? I guess eval mode will be desired in most cases.

Comment (jpcbertoldo) [reply anomalib/models/dfkde/torch_model.py:52]: My personal preferences for this is to not assume much and leave it to the user adding option -- I know, it would be yet a new param...

There are many papers on feature adaptation showing up recently and I think there is no consensus on rather to update the feature extractor or not, so IMO it's hard to assume a better default.

**Review by samet-akcay — CHANGES_REQUESTED** [review]
Similar to what @djdameln raised, my primary concern is the following line in `torch_model.py` and `lightning_model.py`:

```python
feature_extractor: FeatureExtractorParams
```

`feaature_extractor` variable actually contains feature extractor parameters. Storing these as feature_extractor causes a confusion since the codebase has so many feature_extractor here and there.

**Review by djdameln — COMMENTED** [review]
Thanks, huge improvement now that one layer of wrapper has been removed. I am not 100% convinced that we need the wrapper design at all, since the functionality added by the wrapper classes is only used for the construction of the model. But to speed things up, let's stick with this design for now. We can revisit the design once we deprecate timm.

File: anomalib/models/components/feature_extractors/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        Using Backbone params
+
+        >>> from anomalib.models.components.feature_extractors import TorchFXFeatureExtractorParams
+        >>> from torchvision.models.efficientnet import EfficientNet_B5_Weights
+        >>> params = TorchFXFeatureExtractorParams(backbone="efficientnet_b5",
+        ...                     return_nodes=["features.6.8"],
+        ...                     weights=EfficientNet_B5_Weights.DEFAULT
+        ... )
+        >>> FeatureExtractor(params)
+        FeatureExtractor(
```
[the remark this answers is not in the document]
Comment (ashwinvaidya17) [reply anomalib/models/components/feature_extractors/wrapper.py:56]: I need to update the example from `FeatureExtractor` to `get_feature_extractor` but you can directly pass the params without wrapping them in the params class. See this test

https://github.com/openvinotoolkit/anomalib/blob/7f1a42ccbacba154ba77cb1fb1b589202080c91e/tests/pre_merge/models/test_feature_extractor.py#L44

**Comment by jpcbertoldo** [comment]
Is there a github discussion/issue from which one can read why

```
"TimmFeatureExtractor will be removed in the future version."
```

?

**Comment by samet-akcay** [comment]
> Is there a github discussion/issue from which one can read why
> 
> ```
> "TimmFeatureExtractor will be removed in the future version."
> ```
> 
> ?

@jpcbertoldo, we have recently spotted that timm feature extractor provides inconsistent results, which overall adversely impact the performance of the models. We therefore decided to use torchfx feature extractor

**Comment by ashwinvaidya17** [comment]
> Is there a github discussion/issue from which one can read why
> 
> ```
> "TimmFeatureExtractor will be removed in the future version."
> ```
> 
> ?

I've gathered our motivation here https://github.com/openvinotoolkit/anomalib/discussions/779

File: anomalib/models/components/feature_extractors/__init__.py [hunk 7]
```diff
@@ -3,14 +3,16 @@
 # Copyright (C) 2022 Intel Corporation
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/__init__.py:3]: Should this module be called `feature_extraction`? This would be more in line other modules under `models.components` (e.g. `dimensionality_reduction` and `sampling`)

File: anomalib/models/components/feature_extractors/utils.py [hunk 8]
```diff
@@ -1,27 +1,126 @@
-"""Utility functions to manipulate feature extractors."""
+"""Convenience wrapper for feature extraction methods."""
```
Comment (djdameln) [inline anomalib/models/components/feature_extractors/utils.py:1]:
```suggestion
"""Utility functions for parsing feature extraction parameters."""
```

**Review by djdameln — COMMENTED** [review]
Thanks. Just two minor comments left:

File: anomalib/models/stfpm/lightning_model.py [hunk 9]
```diff
@@ -27,22 +28,19 @@ class Stfpm(AnomalyModule):
 
     Args:
         input_size (Tuple[int, int]): Size of the model input.
-        backbone (str): Backbone CNN network
-        layers (List[str]): Layers to extract features from the backbone CNN
+        student_teacher_model (FeatureExtractorParams): Teacher model parameters.
```
Comment (samet-akcay) [inline anomalib/models/stfpm/lightning_model.py:31]: Can you remind me the main motivation of naming this as `student_teacher_model`? Can't this be `teacher_model`?

**Review by samet-akcay — COMMENTED** [review]
Thanks, only a single comment. Thanks a lot for your effort!

File: anomalib/models/stfpm/lightning_model.py (hunk 9 above)
Comment (ashwinvaidya17) [reply anomalib/models/stfpm/lightning_model.py:31]: Looks like I forgot to update the description of this argument, but the idea is that since we use the same backbone for both student and teacher model, the argument should indicate the same.

**Review by samet-akcay — APPROVED** [review]

**Review by djdameln — APPROVED** [review]
Thanks!
